File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> inward(numCourses,0); 
        vector<bool> visited(numCourses,false); 
        vector<vector<int>> adj(numCourses); 
        for(vector<int>& v : prerequisites){
            int to = v[0], from = v[1]; 
            inward[to]++; 
            adj[from].push_back(to); 
        }

        queue<int> q; 
        for(int i = 0; i < numCourses; i++){
            if(inward[i] == 0) q.push(i); 
        }

        if(q.empty()) return false; //there is a cycle 

        while(!q.empty()){
            int size = q.size(); 
            for(int i = 0; i < size; i++){
                int node = q.front(); 
                q.pop(); 
                visited[node] = true; 
                for(int next : adj[node]){
                    inward[next]--; 
                    if(inward[next] <= 0){
                        q.push(next); 
                    }
                }
            }
        }

        for(int i = 0; i < numCourses; i++){
            if(!visited[i]) return false; 
        }

        return true; 

    }
};

File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[] inOrder = new int[numCourses]; 
        List<List<Integer>> adj = new ArrayList<>(); 
        for(int i = 0; i < numCourses; i++){
            adj.add(new ArrayList<>()); 
        }
        for(int[] pre : prerequisites){
            int to = pre[0]; 
            int from = pre[1]; 

            inOrder[to]++; 
            adj.get(from).add(to); 
        }

        Queue<Integer> q = new LinkedList<>(); 
        for(int i = 0; i < numCourses; i++){
            if(inOrder[i] == 0) q.add(i); 
        }

        if(q.isEmpty()) return false; 

        int num = 0; 

        while(!q.isEmpty()){
            int size = q.size(); 
            for(int i = 0; i < size; i++){
                int node = q.poll();
                num++; 
                for(int next : adj.get(node)){
                    inOrder[next]--; 
                    if(inOrder[next] == 0){
                        q.add(next); 
                    }
                }
            }
        }

        return num == numCourses; 
    }
}